homeassistant/components/modbus/base_platform.py

```diff
@@ -54,8 +54,8 @@ def __init__(self, hub: ModbusHub, entry: dict[str, Any]) -> None:
         self._device_class = entry.get(CONF_DEVICE_CLASS)
         self._input_type = entry[CONF_INPUT_TYPE]
         self._value = None
-        self._available = True
         self._scan_interval = int(entry[CONF_SCAN_INTERVAL])
+        self._available = self._scan_interval == 0 or self._scan_interval > 1000
```

Comment:
Why should the entity be available when scan interval is longer than ~16 minutes and client not connected.

Reply:
Because when HA starts we do not know if the client is connected. The code in here might not come into play before after 30minutes (when pymodbus gives up).

Reply:
Apart from that setting the entity to false at startup, seems to have broken a lot of the test cases (no scan_interval configured), I will repair those.

Reply:
If the client connect attempt fails we know we the client isn't connected. So until the connection succeeds the entities should be unavailable, right?

Reply:
test cases corrected.

Reply:
that would be half cooked, Example a device connects and is made available with the first update...when it later disconnects (for whatever reason) we are not notified, but simply see the update failing and hence sets the device to unavailable (reconnect happens automatically in pymodbus, without us knowing about it).

You want to make a special case of a very special situation, the device does not connect during startup (and only startup) you want it marked as unavailable, meaning we never try to do updates (or service calls) so we never detect if the device connect, and the user must wait the full 10 minutes until next connect retry. I do not see this as user friendly.

Even if the connect() fail "normally" our retry code will not be activated. Our retry code is for more severe cases like when you pull the rj45.

Reply:
The scheduled updates aren't turned off when the entity is unavailable.

Reply:
No but the call to pymodbus is blocked. line 317:
if self._config_delay:
            return None

However I promised myself a while ago not to discuss, so the retry mechanism for the special startup cases are removed. The rest of the PR are still valid since it leaves it up to pymodbus to do retry.

The automatic retry in pymodbus are very much needed, since it is likely a device sooner or later will reboot etc, the only drawback is that if one of the special cases are hit during startup, a HA reboot is needed.

Reply:
What's the relationship between `self._config_delay` and `self._scan_interval`? I don't understand why there's a special case for scan interval longer than 1000 seconds.

Reply:
It is explained in the documentation. self._config_delay is a delay imposed after the first connect until the first request is sent. This is needed because some devices need time to prepare, so if you issue a request directly after connect the device abort the request.

self._scan_interval, is the frequency used to request data after the first request.

Long time ago, scan_interval: needed to be > 0, and as I explained earlier some entities like e.g. a switch (if used without a physical switch) does not need to be polled, because the state will only change after a command. Due to that fact people started configure a very high scan_interval (suggested at that time was 9999) to avoid requesting unneeded data. Some releases ago I added the possibility to set scan_interval to 0, meaning no poll at all. Important for both cases are that the e.g. switch needs to be available otherwise you cannot do a service call. 

I have removed the availability changes, so now the code is back to how it was before this PR. An entity is available until the first failing update call. I have also removed the changed test files, because even though it makes sense, the change is not needed for this PR.

Now we are down to the very much needed change that allow pymodbus (just see the number of issues this PR solves) to reconnect after the system is started, and problems during startup (earlier special cases) must be solved with a HA reboot.